snuba/query/processors/uuid_column_processor.py

```diff
@@ -26,6 +27,141 @@
 metrics = MetricsWrapper(environment.metrics, "api.query.uuid_processor")
 
 
+class UUIDTranslationError(Exception):
+    pass
+
+
+class BetterUUIDColumnProcessor(QueryProcessor):
+    """
+    Processor that handles columns which are stored as UUID in ClickHouse but
+    are represented externally as a non hyphenated hex string.
+    """
+
+    def __init__(self, uuid_columns: Set[str]) -> None:
+        self.__uuid_columns = uuid_columns
+        uuid_column_match = Or([String(col) for col in uuid_columns])
+
+        operator = Param(
+            "operator",
+            Or(
+                [
+                    String(op)
+                    for op in FUNCTION_TO_OPERATOR
+                    if op not in (ConditionFunctions.IN, ConditionFunctions.NOT_IN)
+                ]
+            ),
+        )
+
+        in_operators = Param(
+            "operator",
+            Or((String(ConditionFunctions.IN), String(ConditionFunctions.NOT_IN))),
+        )
+
+        col_func = FunctionCallMatch(
+            String("replaceAll"),
+            (
+                FunctionCallMatch(
+                    String("toString"),
+                    (Param("col", ColumnMatch(None, uuid_column_match)),),
+                ),
+            ),
+            with_optionals=True,
+        )
+
+        literal = Param("literal", LiteralMatch(AnyMatch(str)))
+
+        self.__condition_matcher = Or(
+            [
+                FunctionCallMatch(operator, (literal, col_func)),
+                FunctionCallMatch(operator, (col_func, literal)),
+            ]
+        )
+
+        self.__in_condition_matcher = FunctionCallMatch(
+            in_operators,
+            (col_func, Param("tuple", FunctionCallMatch(String("tuple"), None)),),
+        )
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        def translate_uuid_literal(exp: Expression) -> Expression:
+            try:
+                assert isinstance(exp, Literal)
+                new_val = str(uuid.UUID(str(exp.value)))
+                return Literal(alias=exp.alias, value=new_val)
+            except (AssertionError, ValueError):
+                raise UUIDTranslationError("Not a valid UUID string")
+
+        def process_condition(exp: Expression) -> Expression:
+            match = self.__condition_matcher.match(exp)
+
+            if match:
+                try:
+                    return FunctionCall(
+                        exp.alias,
+                        match.string("operator"),
+                        (
+                            match.expression("col"),
+                            translate_uuid_literal(match.expression("literal")),
+                        ),
+                    )
+                except UUIDTranslationError:
+                    # We probably won't get to this point since we should have validations earlier
```

Comment:
Why wouldn't we return the original expression? It's probably wrong, but we shouldn't hide the error message if the client is sending bad queries.

Reply:
The original expression would cause an error on ClickHouse, and there's no point running a query that errors. We can raise a more meaningful error here if we wanted to. But the reason I didn't was because I think the UUID storage is just a Snuba detail. As far as Sentry is concerned it's just a string. In the case of event ID where this is currently used a string that isn't a proper UUID would only error on the errors table (which stores that column as UUID) and not the events table which stored as string. So allowing the query to run successfully by mapping this to 0 is more compatible with the previous result we got from the events table.

Reply:
I suppose that makes sense. Let's leave this then. It still bums me out that this isn't an error.